### PR TITLE
Upstream tracing APIs in a generalized way

### PR DESCRIPTION
**Pull request by ZacSweers**

### CODE REVIEW

**Review by arpita184 — APPROVED** [review]
Few Nits

File: tracing/src/main/kotlin/com/slack/sgp/tracing/reporter/SimpleTraceReporter.kt [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+public class SimpleTraceReporter(
+  private val endpoint: String,
+  private val client: Lazy<OkHttpClient>,
+) : TraceReporter {
+
+  private val tracingService =
+    Retrofit.Builder()
+      .callFactory { client.value.newCall(it) }
+      .addConverterFactory(WireConverterFactory.create())
+      .baseUrl("https://example.com") // Throwaway, URL is always passed directly in
```
[the remark this answers is not in the document]
Comment (ZacSweers) [reply tracing/src/main/kotlin/com/slack/sgp/tracing/reporter/SimpleTraceReporter.kt:35]: If we did that, we'd have to construct a new Retrofit instance every time a call is made. With this approach we can keep the same instance and pass the url on directly.

File: tracing/src/main/kotlin/com/slack/sgp/tracing/model/Spans.kt [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+public fun buildSpan(
+  name: String,
+  startTimestampMicros: Long,
+  durationMicros: Long,
+  traceId: ByteString = makeId(),
+  parentId: ByteString = ByteString.EMPTY,
+  addTags: TagBuilder.() -> Unit = {},
+): Span {
+  return Span(
+    id = makeId(),
```
[the remark this answers is not in the document]
Comment (ZacSweers) [reply tracing/src/main/kotlin/com/slack/sgp/tracing/model/Spans.kt:32]: It's the ID of this span. You can have multiple spans within a single trace

[the remark this answers is not in the document]
Comment (arpita184) [reply tracing/src/main/kotlin/com/slack/sgp/tracing/model/Spans.kt:32]: Ignore this comment. Later on this made sense but just forgot to remove it :)

File: tracing/build.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,24 @@
+plugins {
+  kotlin("jvm")
+  alias(libs.plugins.wire)
+  alias(libs.plugins.mavenPublish)
+}
+
+wire {
```
[the remark this answers is not in the document]
Comment (arpita184) [reply tracing/build.gradle.kts:7]: There was a typo. What I want was this is not so common configuration so adding a comment might be helpful.

[the remark this answers is not in the document]
Comment (ZacSweers) [reply tracing/build.gradle.kts:7]: This is pretty by the book from the docs you linked, is it not self-explanatory? 🤔

**Pull request merged**